### PR TITLE
Add toString() support for all datum types #1553

### PR DESCRIPTION
**Pull request by xd1313113**
## Relevant Issues
- Closes Issue #1553 

## Description
- Add toString() support for all datum types
- Make toString consistent

## Other Information
- Updated Unreleased Section in CHANGELOG: **YES**
- Any backward-incompatible changes? **NO**
- Any new external dependencies? **YES**  Add java Snapshot test and its dependency slf4j
- Do your changes comply with the [contributing][cg] and [code style][csg] guidelines? **YES**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

[cg]: https://github.com/partiql/partiql-lang-kotlin/blob/main/CONTRIBUTING.md
[csg]: https://github.com/partiql/partiql-lang-kotlin/blob/main/CODE_STYLE.md

### CODE REVIEW

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumRow.java [hunk 1]
```diff
@@ -4,11 +4,7 @@
 import org.partiql.spi.types.PType;
 import org.partiql.spi.types.PTypeField;
 
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (XuechunHHH) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumRow.java]: Shall we avoid wildcard import here according to the code style? Similar in other files.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java [hunk 2]
```diff
@@ -82,15 +78,25 @@ public PType getType() {
...
+            StringJoiner innerJoiner = new StringJoiner(", ", "[", "]");
+            for (Datum d : value) {
+                innerJoiner.add(d.toString());
+            }
+            sb.append(innerJoiner);
+
+            outerJoiner.add(sb.toString());
+        });
+
+        return "DatumStruct{" +
```
Comment (XuechunHHH) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java:81]: Can we create a helper function like:
```
public static String datumToString(String className, Object type, Object value) {
    return "Datum" + className + "{_type=" + type + ", _value=" + value + '}';
}
```
to avoid the code repetition?

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumBytes.java [hunk 3]
```diff
@@ -34,4 +34,17 @@ public byte[] getBytes() {
     public PType getType() {
         return _type;
     }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
+        for (byte b : _value) {
+            sb.append(String.format("%02X", b & 0xFF));
```
Comment (alancai98) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumBytes.java]: mind clarifying what this printing logic is doing? I think it may be better to just leave it as bytes since the primary use case of `toString` is for our own debugging.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumCollection.java [hunk 4]
```diff
@@ -39,9 +42,13 @@ public PType getType() {
...
+        StringJoiner joiner = new StringJoiner(", ", "[", "]");
+
+        _value.forEach(d -> joiner.add(d.toString()));
+
         return "DatumCollection{" +
                 "_type=" + _type +
-                ", _value=" + _value +
+                ", _value=" + joiner +
                 '}';
     }
```
Comment (alancai98) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumCollection.java]: Doing a `+` here will create a new immutable string, which can be a redundant operation if the collection values' string representation is large. For any container datums (e.g. collection, row, struct), we should try reducing the `+` usage in favor of `StringBuilder`s. For scalar values, they tend to be a small, fixed size so it's not too necessary to migrate to use `StringBuilder`. I'll leave it up to you if you want to migrate those other classes.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java [hunk 5]
```diff
@@ -123,6 +123,9 @@ public int getNanos() {
 
     @Override
     public String toString() {
-        return "INTERVAL '" + days + " " + hours + ":" + minutes + ":" + seconds + "." + nanos + "' " + getType();
+        return "DatumIntervalDayTime{" +
```
Comment (alancai98) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java:126]: Missing a few fields like interval code, precision, and fractional precision.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalYearMonth.java [hunk 6]
```diff
@@ -58,8 +58,10 @@ public int getMonths() throws InvalidOperationException, NullPointerException {
         return _months;
     }
 
-    @Override
     public String toString() {
-        return "INTERVAL '" + _years + "-" + _months + "' " + getType();
+        return "DatumIntervalYearMonth{" +
```
Comment (alancai98) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalYearMonth.java:62]: Missing a few fields like precision and interval code

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java [hunk 7]
```diff
@@ -82,15 +78,25 @@ public PType getType() {
...
-        return sb.toString();
+
+            StringJoiner innerJoiner = new StringJoiner(", ", "[", "]");
+            for (Datum d : value) {
+                innerJoiner.add(d.toString());
+            }
+            sb.append(innerJoiner);
+
+            outerJoiner.add(sb.toString());
+        });
```
Comment (alancai98) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java]: This block is the same as what's in `DatumRow`. Let's extract it to a separate method

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumRow.java (hunk 1 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumRow.java]: Yeah let's get rid of the * imports here. Since the files are written in Java, they aren't caught by ktlint

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java [hunk 8]
```diff
@@ -82,15 +78,25 @@ public PType getType() {
...
+            StringBuilder sb = new StringBuilder();
+            sb.append(key);
             sb.append(": ");
-            sb.append(entry.getValue().toString());
-            sb.append(", ");
-        }
-        sb.append(" }");
-        return sb.toString();
+
+            StringJoiner innerJoiner = new StringJoiner(", ", "[", "]");
```
Comment (alancai98) [inline partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java]: I don't think we need to wrap these in the square brackets, `[` & `]`.

File: partiql-spi/src/test/kotlin/org/partiql/spi/value/DatumToStringTest.kt [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+import java.time.ZoneOffset
+import kotlin.test.assertEquals
+
+/**
+ * Test cases for toString methods of all Datum implementations.
+ * Verifies that each Datum class produces expected string representations.
+ *
+ * Test methods are in alphabetical order by method name
+ */
+class DatumToStringTest {
```
Comment (alancai98) [inline partiql-spi/src/test/kotlin/org/partiql/spi/value/DatumToStringTest.kt:34]: While I think unit tests are important to add for most things, I think these hand-written `toString` unit tests may be annoying to maintain over time. For instance, if we added another field to be printed by each `Datum`, we would have to manually update all of the existing tests to match this string representation. IMO this is slight overkill for `toString` since we'll primarily use it for debugging.

This has been my experience especially with our [plan typing tests](https://github.com/partiql/partiql-lang-kotlin/blob/b5be0aa4f6e7c416c8443c78c53c39984ea772aa/partiql-planner/src/test/kotlin/org/partiql/planner/internal/typer/PlanTyperTestsPorted.kt), especially when doing any sort of refactor/migration. Our old evaluation tests had this issue until we transitioned to use the conformance tests, `partiql-tests`.

Perhaps this is a good chance for us to start using snapshot testing -- https://github.com/origin-energy/java-snapshot-testing. I've found it to be a lot simpler to maintain these tests in some other codebases. We already do snapshot testing in our Rust implementation. See here for an example setup using the Java/Kotlin version -- https://github.com/partiql/partiql-graphviz-java/pull/11.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java (hunk 2 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java:81]: IMO, it's not too much repetition and there's enough differences between the `Datum`s. We should refactor some repeated logic we use for struct and row though

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumRow.java (hunk 1 above)
Comment (xd1313113) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumRow.java]: Thanks. Will fix IDE seems to be the evil.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumBytes.java (hunk 3 above)
Comment (xd1313113) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumBytes.java]: Will add comment, this will convert to HEX value. I think byte array is more readable in HEX. The java default print number in radix 10. Do you have preference?

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumCollection.java (hunk 4 above)
Comment (xd1313113) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumCollection.java]: The collection Datums are using stringjoiner to avoid immutable string concatenation for value.
Also I am thinking to String.format to reduce + currently in return statement

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java (hunk 5 above)
Comment (xd1313113) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java:126]: The _type has already those info. So remove it from value.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java (hunk 8 above)
Comment (xd1313113) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java]: Why? The struct is key-value pair and value is List<Datum>. I use "[]" for array/list representation

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java (hunk 5 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java:126]: The `_type` does not include the `intervalCode` since the PType only tells you if it's a year-month interval or a day-time interval. The interval code (https://github.com/partiql/partiql-lang-kotlin/blob/dixiao/daturmtoStr/partiql-spi/src/main/java/org/partiql/spi/types/IntervalCode.java#L16-L28) will actually tell you which of the original fields was specified in the datum. Like whether the year-month interval was just a `year` or just a `month`.

The `PType` does include the `precision` and `fractionalPrecision` printed back as a type (like `DAY(3) TO SECOND (5)`) but perhaps it would be more readable as separate fields. I'm less opinionated on these specific fields.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumBytes.java (hunk 3 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumBytes.java]: IMO, since the `toString` is used for debugging, I think the output should just be as similar to what may be passed in as part of the constructor to a `DatumBytes`. For the example you included in the tests, it's hard to easily understand how `Datum.blob(byteArrayOf(10, 50, 100, 127))` corresponds to the value `0A32647F` without diving into the implementation. I think it would be clearer if the output string representation was a list of the input values `[10, 50, 100, 12]`.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java (hunk 8 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java]: Ah I see. Forgot that `DatumStruct`s and `DatumRows` permit duplicate key names. Perhaps can add a test using the duplicate key names to check that the `innerJoiner` works for > 1 element.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java (hunk 5 above)
Comment (xd1313113) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java:126]: It actually does 
 assertEquals("DatumIntervalYearMonth{**_type=YEAR (2) TO MONTH**, _value=INTERVAL '2-6'}", datum.toString())

Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumIntervalDayTime.java:126]: Ah ok, it wasn't clear to me that the `PType` `toString` encodes the `intervalCode`. Let's add a couple explicit tests that don't include all of the fields. E.g.
- `Datum.intervalYear(10, 2)`
- `Datum.intervalDay(15, 2)`

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumCollection.java (hunk 4 above)
Comment (xd1313113) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumCollection.java]: I took a further look. "+" is fine here as compiler actually compiles + into StringBuilder. String.format is actually more expensive. Collection object are built by Stringjoiner which internally use stringbuiler

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java (hunk 8 above)
Comment (xd1313113) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java]: I took a look that for same key, it get chained in the list of value. duplicate key test is not necessary for toStringTest as the tests focuses on structure of the object, which is covered, flat and nested.

File: partiql-spi/src/test/kotlin/org/partiql/spi/value/DatumToStringTest.kt (hunk 9 above)
Comment (xd1313113) [reply partiql-spi/src/test/kotlin/org/partiql/spi/value/DatumToStringTest.kt:34]: Agree that toString wasn't that important for most use scenarios. my thought is it does not hurt to add it for now as toString is **public** and we customize its implementation and implementation is still under test. For major refactor/migration, they may need to make a decision to keep it, disable or fix later

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumCollection.java (hunk 4 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumCollection.java]: > I took a further look. "+" is fine here as compiler actually compiles + into StringBuilder. String.format is actually more expensive.

Not to disagree with your point, but I'm curious how did you verify this? In the docs or some local testing? My read of this discussion https://stackoverflow.com/questions/1532461/stringbuilder-vs-string-concatenation-in-tostring-in-java is that they `+` and `StringBuilder` `append` are largely equivalent in more recent versions of Java. Seems like the issue w/ combining `String`s is when using `+=` in a loop.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java (hunk 8 above)
Comment (alancai98) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java]: IMO, duplicate keys could still be a good test to add.

File: partiql-spi/src/test/kotlin/org/partiql/spi/value/DatumToStringTest.kt (hunk 9 above)
Comment (alancai98) [reply partiql-spi/src/test/kotlin/org/partiql/spi/value/DatumToStringTest.kt:34]: Could we try using snapshot testing following my other PR to setup (https://github.com/partiql/partiql-graphviz-java/pull/11/files)? A small change to the `toString` such as adding a new property to a `Datum` or modifying the printing of `PType`s could require tedious manual effort in updating the unit test.

It should be easy to set up -- just add a line to the spi build files and annotations to all of the existing tests. It can also give you good practice working with our current project builds (like adding additional dependencies). I can help answer any questions w/ the setup.

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumCollection.java (hunk 4 above)
Comment (xd1313113) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumCollection.java]: yeah, forget to mentioned. I actually changed to StringBuilder, however intellij just remind me not needed as complier will compile static "+" to Stringbuilder. I also did some local test. There seems the performance diff are subtle in my toString implementation. And we only need to avoid += in loop where we use stringjoiner for our collection Datum

File: partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java (hunk 8 above)
Comment (xd1313113) [reply partiql-spi/src/main/java/org/partiql/spi/value/DatumStruct.java]: added

File: partiql-spi/src/test/kotlin/org/partiql/spi/value/DatumToStringTest.kt (hunk 9 above)
Comment (xd1313113) [reply partiql-spi/src/test/kotlin/org/partiql/spi/value/DatumToStringTest.kt:34]: add snapshot test

File: buildSrc/src/main/kotlin/partiql.versions.kt [hunk 10]
```diff
@@ -53,6 +53,8 @@ object Versions {
     const val junit4 = "4.12"
     const val junit4Params = "1.1.1"
     const val mockito = "4.5.0"
+    const val snapshotTest = "4.0.8"
+    const val slf4j = "2.0.17"
```
Comment (alancai98) [inline buildSrc/src/main/kotlin/partiql.versions.kt:57]: Thanks for going about adding the snapshot tests. Can we remove this slf4j dependency? I didn't need to use it in the other PR https://github.com/partiql/partiql-graphviz-java/pull/11.

File: partiql-spi/src/test/kotlin/org/partiql/spi/value/DatumToStringTest.kt [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+import java.time.LocalTime
+import java.time.OffsetDateTime
+import java.time.OffsetTime
+import java.time.ZoneOffset
+
+/**
+ * Test cases for toString methods of all Datum implementations.
+ * Verifies that each Datum class produces expected string representations.
+ *
+ * Test methods are in alphabetical order by method name
```
Comment (alancai98) [inline partiql-spi/src/test/kotlin/org/partiql/spi/value/DatumToStringTest.kt:18]: nit: let's add some brief docs on how to run the snapshot tests and update the results if there's a divergence

**Review by alancai98 — COMMENTED** [review]
A couple last minor comments. Thanks for migrating the tests to snapshot testing

**Comment by alancai98** [comment]
Also the PR description probably needs updated to reflect the updated CHANGELOG and the new testing dependency.

File: buildSrc/src/main/kotlin/partiql.versions.kt (hunk 10 above)
Comment (xd1313113) [reply buildSrc/src/main/kotlin/partiql.versions.kt:57]: It is not optional, per Quick Start section from https://github.com/origin-energy/java-snapshot-testing

If removed, follow error is thrown
org.slf4j.LoggerFactory
java.lang.ClassNotFoundException: org.slf4j.LoggerFactory
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	at au.com.origin.snapshots.junit5.SnapshotExtension.<clinit>

File: partiql-spi/src/test/kotlin/org/partiql/spi/value/DatumToStringTest.kt (hunk 11 above)
Comment (xd1313113) [reply partiql-spi/src/test/kotlin/org/partiql/spi/value/DatumToStringTest.kt:18]: added
